### PR TITLE
[Inductor] Add SDPA pattern for OOB GPT2 models

### PR DESCRIPTION
**Pull request by Valentine233**
Add SDPA pattern for 2 OOB models:
- token-classification+gpt2
- text-generation+gpt2

Note that these models have two masks: attention mask with float type and causal mask with bool type.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: torch/_inductor/fx_passes/fuse_attention.py
Comment (kadeng) [inline torch/_inductor/fx_passes/fuse_attention.py:511]: Referring to my comment above: Does this dtype ( in the model we apply this pattern to ) differ from the query tensor dtype in practice?

File: torch/_inductor/fx_passes/fuse_attention.py
Comment (kadeng) [inline torch/_inductor/fx_passes/fuse_attention.py:516]: In the pattern, the causal mask dtype equals the query dtype..

File: torch/_inductor/fx_passes/fuse_attention.py
Comment (kadeng) [inline torch/_inductor/fx_passes/fuse_attention.py:527]: And here in the replacement, the causal mask dtype also equals the query dtype. Is the change in attention.cpp really neccessary?

**Comment by kadeng** [comment]
Looks good, thanks for your contribution. 

I would like to make sure we only make neccessary changes, and that this does not lead to any performance regressions. So I will add the neccessary tags to run this through the torch benchmark suite. Until then I would hold off an approval.

File: aten/src/ATen/native/transformers/attention.cpp
Comment (Valentine233) [reply aten/src/ATen/native/transformers/attention.cpp:500]: For the OOB GPT2 models, we find that `attn_mask` is always float no matter what query is. You could refer [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_gpt2.py#L310).

File: torch/_inductor/fx_passes/fuse_attention.py
Comment (Valentine233) [reply torch/_inductor/fx_passes/fuse_attention.py:527]: Causal mask is bool and attention mask is float. Here is the fx_graph_readable which provides the types. You could see the line of `add_2`: `attn_weights` changes from BF16 to FP32 due to the addition with attention mask (float).

```
# File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:182 in _attn, code: attn_weights = torch.matmul(query, key.transpose(-1, -2))
        permute_3: "bf16[1, 12, 64, 23]" = torch.ops.aten.permute.default(permute_1, [0, 1, 3, 2])
        expand: "bf16[1, 12, 23, 64]" = torch.ops.aten.expand.default(permute, [1, 12, 23, 64]);  permute = None
        view_5: "bf16[12, 23, 64]" = torch.ops.aten.reshape.default(expand, [12, 23, 64]);  expand = None
        expand_1: "bf16[1, 12, 64, 23]" = torch.ops.aten.expand.default(permute_3, [1, 12, 64, 23]);  permute_3 = None
        view_6: "bf16[12, 64, 23]" = torch.ops.aten.reshape.default(expand_1, [12, 64, 23]);  expand_1 = None
        bmm: "bf16[12, 23, 23]" = torch.ops.aten.bmm.default(view_5, view_6);  view_5 = view_6 = None
        view_7: "bf16[1, 12, 23, 23]" = torch.ops.aten.reshape.default(bmm, [1, 12, 23, 23]);  bmm = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:185 in _attn, code: attn_weights = attn_weights / torch.full(
        full_default: "bf16[]" = torch.ops.aten.full.default([], 8.0, dtype = torch.bfloat16, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
        div: "bf16[1, 12, 23, 23]" = torch.ops.aten.div.Tensor(view_7, full_default);  view_7 = full_default = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:196 in _attn, code: causal_mask = self.bias[:, :, key_length - query_length : key_length, :key_length].to(torch.bool)
        _frozen_param15: "b8[1, 1, 23, 23]" = self._frozen_param15
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:200 in _attn, code: mask_value = torch.full([], mask_value, dtype=attn_weights.dtype).to(attn_weights.device)
        full_default_1: "bf16[]" = torch.ops.aten.full.default([], -3.3895313892515355e+38, dtype = torch.bfloat16, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:201 in _attn, code: attn_weights = torch.where(causal_mask, attn_weights.to(attn_weights.dtype), mask_value)
        where: "bf16[1, 12, 23, 23]" = torch.ops.aten.where.self(_frozen_param15, div, full_default_1);  _frozen_param15 = div = full_default_1 = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:205 in _attn, code: attn_weights = attn_weights + attention_mask
        add_2: "f32[1, 12, 23, 23]" = torch.ops.aten.add.Tensor(where, arg14_1);  where = arg14_1 = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:207 in _attn, code: attn_weights = nn.functional.softmax(attn_weights, dim=-1)
        amax: "f32[1, 12, 23, 1]" = torch.ops.aten.amax.default(add_2, [-1], True)
        sub_1: "f32[1, 12, 23, 23]" = torch.ops.aten.sub.Tensor(add_2, amax);  add_2 = amax = None
        exp: "f32[1, 12, 23, 23]" = torch.ops.aten.exp.default(sub_1);  sub_1 = None
        sum_1: "f32[1, 12, 23, 1]" = torch.ops.aten.sum.dim_IntList(exp, [-1], True)
        div_1: "f32[1, 12, 23, 23]" = torch.ops.aten.div.Tensor(exp, sum_1);  exp = sum_1 = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:210 in _attn, code: attn_weights = attn_weights.type(value.dtype)
        convert_element_type_9: "bf16[1, 12, 23, 23]" = torch.ops.prims.convert_element_type.default(div_1, torch.bfloat16);  div_1 = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:211 in _attn, code: attn_weights = self.attn_dropout(attn_weights)
        clone: "bf16[1, 12, 23, 23]" = torch.ops.aten.clone.default(convert_element_type_9);  convert_element_type_9 = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:217 in _attn, code: attn_output = torch.matmul(attn_weights, value)
        expand_2: "bf16[1, 12, 23, 23]" = torch.ops.aten.expand.default(clone, [1, 12, 23, 23]);  clone = None
        view_8: "bf16[12, 23, 23]" = torch.ops.aten.reshape.default(expand_2, [12, 23, 23]);  expand_2 = None
        expand_3: "bf16[1, 12, 23, 64]" = torch.ops.aten.expand.default(permute_2, [1, 12, 23, 64])
        view_9: "bf16[12, 23, 64]" = torch.ops.aten.reshape.default(expand_3, [12, 23, 64]);  expand_3 = None
        bmm_1: "bf16[12, 23, 64]" = torch.ops.aten.bmm.default(view_8, view_9);  view_8 = view_9 = None
        view_10: "bf16[1, 12, 23, 64]" = torch.ops.aten.reshape.default(bmm_1, [1, 12, 23, 64]);  bmm_1 = None
```

File: torch/_inductor/fx_passes/fuse_attention.py
Comment (Valentine233) [reply torch/_inductor/fx_passes/fuse_attention.py:516]: `causal_mask_value` uses the same type as query, but causal mask is bool.

**Comment by Valentine233** [comment]
> Looks good, thanks for your contribution.
> 
> I would like to make sure we only make neccessary changes, and that this does not lead to any performance regressions. So I will run the benchmark suite. Until then I would hold off an approval.
> 
> The benchmark run I triggered can be monitored here: https://github.com/pytorch/pytorch/actions/runs/8967381583

Thanks for your review, @kadeng . The benchmark has 1 successful job and 2 skipped ones. Is this expected?

**Review by jgong5 — APPROVED** [review]

**Review by kadeng — APPROVED** [review]
Torchinductor Performance dashboard (results of the benchmark runs, see https://hud.pytorch.org/benchmark/compilers and set one of the branches to sdpa_gpt2) looks good as well. There are even some training speedups visible, but I'm not sure if these are really attributable to this diff.

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
@Valentine233 can you add a `@skipIfRocm` to the failing test?

**Comment by Valentine233** [comment]
> @Valentine233 can you add a `@skipIfRocm` to the failing test?

Sure! Added for `TestFullyShard2DTraining(FSDPTest) - test_train_parity_2d_mlp`.

**Review by jansel — APPROVED** [review]

**Comment by Valentine233** [comment]
@kadeng @jansel `Meta Internal-Only Changes Check` failed due to the code change. Could you help import the Pull Request? Thanks~

**Comment by jansel** [comment]
> @kadeng has imported this pull request. If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D57045736).

It is because of this it is now linked to an internal diff and cant be landed by pytorchbot.

**Pull request closed without merging**
